### ISSUE TITLE
nerfacto checkpoints render fail with latest 

### ISSUE THREAD

**Issue by machenmusik** [nerfstudio-project/nerfstudio#1523]
Last Friday, I was able to use ns-render with my trained nerfacto model (with commit with 3933589).
This morning the latest gives this error when trying to render
```
RuntimeError: Error(s) in loading state_dict for NerfactoModel:
        Missing key(s) in state_dict: "field.max_res", "field.num_levels", "field.log2_hashmap_size",
"proposal_networks.0.max_res", "proposal_networks.0.num_levels", "proposal_networks.0.log2_hashmap_size",
"proposal_networks.1.max_res", "proposal_networks.1.num_levels", "proposal_networks.1.log2_hashmap_size".
```

**Comment by machenmusik** [comment]
ns-render works with 871eb0f7d3232927b05b6d366998ea5d1f0885b6

**Comment by machenmusik** [comment]
My personal sense is that adding volinga should not break prior nerfacto checkpoints, but others should probably comment as well.  

My understanding was that volinga is merely exposing these keys from nerfacto, so perhaps applying the correct values into the model state when not present for prior nerfacto model configs could solve the issue.

**Comment by Frivas97** [comment]
I think the problem is in base_pipeline.py file in function load_pipeline:

https://github.com/nerfstudio-project/nerfstudio/blob/5d640bfcfdf174922687c11a629a9eb7659a47ce/nerfstudio/pipelines/base_pipeline.py#L373

setting strict=False would solve the problem, but not sure if this won't break something.

**Comment by machenmusik** [comment]
(Yes - although I assume strict=True is there intentionally to make sure the config is for the correct model.)

**Comment by Frivas97** [comment]
Just to confirm the reason, were you able to see new models (with the new keys) in the viewer?

**Comment by machenmusik** [comment]
I can see a few different strategies, including these two
1. No backward compatibility for changes to individual models - for example, prior 0.1.18 nerfacto checkpoints may not work with 0.1.19
2. Specific accommodations for backward compatibility left to individual models - for example, nerfacto model may choose to be non-strict, or to fill in for certain missing parameters

Generally, I think this project has been following strategy 1, but has been careful to minimize changes that would break compatibility (e.g. predict_normals).

The wrinkle here is one model "changed" to accommodate a new variant (here, volinga exposing parameters from nerfacto, and the changes were requested to backport into nerfacto).  

It is pretty unfortunate that all prior nerfacto checkpoints are invalidated even if one doesn't use volinga - and since volinga is not generally available, at the moment, almost every nerfstudio user is broken for nothing.

An example of changes to implement strategy 2 would be something like this, but as a method called on the model to adjust the dict before loading it, e.g. adjust_model_dict within NerfactoModel instead of jammed into base_pipeline.py
```
--- a/nerfstudio/pipelines/base_pipeline.py
+++ b/nerfstudio/pipelines/base_pipeline.py
@@ -45,6 +45,7 @@ from nerfstudio.data.datamanagers.base_datamanager import (
 )
 from nerfstudio.engine.callbacks import TrainingCallback, TrainingCallbackAttributes
 from nerfstudio.models.base_model import Model, ModelConfig
+from nerfstudio.models.nerfacto import NerfactoModel
 from nerfstudio.utils import profiler


@@ -110,6 +111,17 @@ class Pipeline(nn.Module):
             key.replace("_model.", ""): value for key, value in state_dict.items() if key.startswith("_model.")
         }
         pipeline_state = {key: value for key, value in state_dict.items() if not key.startswith("_model.")}
+        # Targeted fix for missing additions to nerfacto model in 9053643932c1ab6fd263fe91eae38c999c2b149f
+        if isinstance(self._model, NerfactoModel):
+            for prefix in ["field", "proposal_networks.0", "proposal_networks.1"]:
+                for key, value in {
+                    "max_res": self._model.field.max_res,
+                    "num_levels": self._model.field.num_levels,
+                    "log2_hashmap_size": self._model.field.log2_hashmap_size,
+                }.items():
+                    fieldkey = prefix + "." + key
+                    if not fieldkey in model_state:
+                        model_state[fieldkey] = torch.tensor(value)
         self._model.load_state_dict(model_state, strict=strict)
         super().load_state_dict(pipeline_state, strict=False)
```

**Comment by machenmusik** [comment]
> Just to confirm the reason, were you able to see new models (with the new keys) in the viewer?
I have not tried, that could take a while; perhaps you can quickly check against nerfacto with one of your datasets?

**Comment by Frivas97** [comment]
I completely agree with you @machenmusik. I would like to know the opinion of @tancik. If he agrees I can implement these changes.

**Comment by machenmusik** [comment]
(Upon closer inspection of the volinga changes, I wouldn't recommend using that code snippet I pasted above; although that does allow prior models to load, it looks like the Nerfacto model would use different default values for "field", "proposal_networks.0", and "proposal_networks.1" )

**Comment by machenmusik** [comment]
> ValueError: loaded state dict contains a parameter group that doesn't match the size of optimizer's group

Perhaps more concerning, I get that same error when doing `ns-train nerfacto --viewer.start-train False ` with a prior Nerfacto model even if using strict=False as discussed at https://github.com/nerfstudio-project/nerfstudio/issues/1523#issuecomment-1446315042

**Comment by machenmusik** [comment]
> were you able to see new models (with the new keys) in the viewer?

Newly trained models seem to behave properly.

**Comment by machenmusik** [comment]
Re: https://github.com/nerfstudio-project/nerfstudio/issues/1523#issuecomment-1446643458

@tancik thoughts?

**Comment by tancik** [comment]
I'm conflicted. I am not a fan of the proposal in https://github.com/nerfstudio-project/nerfstudio/issues/1523#issuecomment-1446875464, I want to avoid having logic related to specific models in the base modules such as Pipeline. While it is not ideal that the checkpoints break between different nerfstudio versions, I don't view it as too much of an issue since training the models is relatively cheap compared to other ml models. Longer term, I don't think we want this, but since the project is still pretty new evolving quickly, I think it's fine to have some breaking changes. One thing we could think about is improving the error messaging around failures like this. Ie, if the state doesn't match we can include a warning "Checkpoint can not be loaded, this could be the result of training with a different version of nerfstudio."

**Comment by machenmusik** [comment]
Agreed that model specific logic should stay in the models.

My current preference would be second option in https://github.com/nerfstudio-project/nerfstudio/issues/1523#issuecomment-1446374171

i.e. give models hooks to provide backward compatibility when loading if they want

Then if volinga is willing to make changes to be backward compatible with prior nerfacto usage, that would be an example others could follow.
